### PR TITLE
Feat(eos_cli_config_gen): Add support for multiple dot1x interface features

### PR DESCRIPTION
**Pull request by mthiel117**
## Change Summary

Adding multiple dot1x features to support the following cli lines:

```
dot1x authentication failure action traffic allow vlan 800
dot1x host-mode multi-host authenticated
dot1x mac based authentication
dot1x timeout tx-period 3
dot1x timeout reauth-period server
dot1x reauthorization request limit 3
```

## Related Issue(s)

n/a

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes
Data Model additions are as follows:
```
dot1x:
  authentication_failure:
    action: < "allow" | "drop" >
    allow_vlan: < 1-4094 >
  host_mode: 
    mode: < "multi-host" | "single-host" >
    multi_host_authenticated: < true | false >
  mac_based_authentication: 
    enabled: < true | false >
    always: < true | false >
    host_mode_common: < true | false >
  timeout:
    idle_host: < 10-65535 >
    quiet_period: < 1-65535 >
    reauth_period: < 60-4294967295 | server >
    reauth_timeout_ignore: < true | false >
    tx_period: < 1-65535 >
  reauthorization_request_limit: < 1-10 >
```

## How to test
Will add molecule test ethernet-interfaces and run the following test:
```
molecule converge -s eos_cli_config_gen -- --limit ethernet-interfaces
```

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

**Comment by mthiel117** [comment]
Looking for approval on this data model before coding the Jinja templates.

**Comment by ClausHolbechArista** [comment]
@mthiel117 looks fine.

**Comment by mthiel117** [comment]
Thanks @ClausHolbechArista 

I will get started on this next week.

**Comment by mthiel117** [comment]
Since dot1x has so many features, it will be hard to note all possible features and values in a markdown table for documentation.  Any ideas on how to handle this?

**Comment by ClausHolbechArista** [comment]
> Since dot1x has so many features, it will be hard to note all possible features and values in a markdown table for documentation. Any ideas on how to handle this?

If the features are in the "nerd knob" category, you can skip the table for those, but for widely used features, we should include them in an overview.

**Comment by mthiel117** [comment]
Looking forward to feedback/suggestions.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2 [hunk 1]
```diff
@@ -209,6 +217,49 @@ interface {{ ethernet_interface.name }}
...
+{%                 if ethernet_interface.dot1x.mac_based_authentication.enabled is arista.avd.defined(true) and
+                      ethernet_interface.dot1x.mac_based_authentication.always is arista.avd.defined(true) %}
+   dot1x mac based authentication always
+{%                 endif %}
+{%                 if ethernet_interface.dot1x.mac_based_authentication.enabled is arista.avd.defined(true) and
+                      ethernet_interface.dot1x.mac_based_authentication.always is not arista.avd.defined and
+                      ethernet_interface.dot1x.mac_based_authentication.host_mode_common is not arista.avd.defined %}
+   dot1x mac based authentication
+{%                 endif %}
+{%             endif %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2]:
```suggestion
{%             if ethernet_interface.dot1x.mac_based_authentication.enabled is arista.avd.defined(true) %}
{%                 set auth_cli = "dot1x mac based authentication" %}
{%                 if ethernet_interface.dot1x.mac_based_authentication.host_mode_common is arista.avd.defined(true) %}
   {{ auth_cli }} host-mode common
{%                 endif %}
{%                 if ethernet_interface.dot1x.mac_based_authentication.always is arista.avd.defined(true) %}
{%                     set auth_cli = auth_cli ~ " always" %}
{%                 endif %}
   {{ auth_cli }}
{%             endif %}
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2 [hunk 2]
```diff
@@ -209,6 +217,49 @@ interface {{ ethernet_interface.name }}
 {%             elif ethernet_interface.dot1x.port_control_force_authorized_phone is arista.avd.defined(false) %}
    no dot1x port-control force-authorized phone
 {%             endif %}
+{%             if ethernet_interface.dot1x.host_mode is arista.avd.defined %}
+{%                 if ethernet_interface.dot1x.host_mode.mode is arista.avd.defined('single-host') %}
+   dot1x host-mode single-host
+{%                 elif ethernet_interface.dot1x.host_mode.mode is arista.avd.defined('multi-host') and
+                        ethernet_interface.dot1x.host_mode.multi_host_authenticated is arista.avd.defined(true) %}
+   dot1x host-mode multi-host authenticated
+{%                 endif %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2]:
```suggestion
{%                 elif ethernet_interface.dot1x.host_mode.mode is arista.avd.defined('multi-host') %}
{%                     set host_mode_cli = "dot1x host-mode multi-host" %}
{%                     if ethernet_interface.dot1x.host_mode.multi_host_authenticated is arista.avd.defined(true) %}
{%                         set host_mode_cli = host_mode_cli ~ " authenticated" %}
{%                     endif %}
   {{ host_mode_cli }}
{%                 endif %}
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/dot1x.j2 [hunk 3]
```diff
@@ -13,13 +13,21 @@
...
+| Interface | PAE Mode | State | Phone Force Authorized | Reauthentication | Auth Failure Action | Host Mode | Mac Based Auth |
+| --------- | -------- | ------| ---------------------- | ---------------- | ------------------- | --------- | -------------- |
 {%     for ethernet_interface in ethernet_interfaces_dot1x %}
 {%         set pae_mode = ethernet_interface.dot1x.pae.mode | arista.avd.default('-') %}
+{%         set auth_failure_action = ethernet_interface.dot1x.authentication_failure.action | arista.avd.default('-') %}
+{%         set auth_failure_allow_vlan = ethernet_interface.dot1x.authentication_failure.allow_vlan | arista.avd.default('-') %}
+{%         if ethernet_interface.dot1x.authentication_failure.action is arista.avd.defined('allow') and
+              ethernet_interface.dot1x.authentication_failure.allow_vlan is arista.avd.defined %}
+{%             set auth_failure_action = ethernet_interface.dot1x.authentication_failure.action ~" vlan " ~ethernet_interface.dot1x.authentication_failure.allow_vlan %}
+{%         endif %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/dot1x.j2]:
```suggestion
{%         set auth_failure_action = ethernet_interface.dot1x.authentication_failure.action | arista.avd.default('-') %}
{%         if auth_failure_action == 'allow' and
              ethernet_interface.dot1x.authentication_failure.allow_vlan is arista.avd.defined %}
{%             set auth_failure_action = auth_failure_action ~ " vlan " ~ ethernet_interface.dot1x.authentication_failure.allow_vlan %}
{%         endif %}
```

**Review by ClausHolbechArista — COMMENTED** [review]
A few suggestions and a bugfix

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2 (hunk 2 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2]: this is fixing multi-host without authenticated

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2 [hunk 4]
```diff
@@ -209,6 +217,47 @@ interface {{ ethernet_interface.name }}
...
+{%             if ethernet_interface.dot1x.mac_based_authentication.enabled is arista.avd.defined(true) %}
+{%                 set auth_cli = "dot1x mac based authentication" %}
+{%                 if ethernet_interface.dot1x.mac_based_authentication.host_mode_common is arista.avd.defined(true) %}
+   {{ auth_cli }} host-mode common
+{%                 endif %}
+{%                 if ethernet_interface.dot1x.mac_based_authentication.always is arista.avd.defined(true) %}
+{%                     set auth_cli = auth_cli ~ " always" %}
+{%                 endif %}
+   {{ auth_cli }}
+{%             endif %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2]:
```suggestion
{%             if ethernet_interface.dot1x.mac_based_authentication.enabled is arista.avd.defined(true) %}
{%                 if ethernet_interface.dot1x.mac_based_authentication.host_mode_common is arista.avd.defined(true) %}
   dot1x mac based authentication host-mode common
{%                     if ethernet_interface.dot1x.mac_based_authentication.always is arista.avd.defined(true) %}
   dot1x mac based authentication always
{%                     endif %}
{%                 else %}
{%                     set auth_cli = "dot1x mac based authentication" %}
{%                     if ethernet_interface.dot1x.mac_based_authentication.always is arista.avd.defined(true) %}
{%                         set auth_cli = auth_cli ~ " always" %}
{%                     endif %}
   {{ auth_cli }}
{%                 endif %}
{%             endif %}
```

**Review by ClausHolbechArista — APPROVED** [review]
LGTM. Verified CLI output on a DUT, and verified order of commands too.

**Review by tgodaA — APPROVED** [review]
LGTM

**Pull request merged**